### DISCUSSION TITLE
Welcome to pedal Discussions!

### DISCUSSION THREAD

**Discussion by acbart** [pedal-edu/pedal#73] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by jlehenbauer** [comment]
This may be out of the purview of this project, but Pedal has ben implemented as a tool to use on the BlockPy editor at [think.cs.vt.edu/blockpy](https://think.cs.vt.edu/blockpy). I'm able to perform `unit_test`s and most of the major functions of Pedal, but can't figure out how to implement the scoring within the site. I'm unable to affect a score or set a pass/fail using `score` inside `unit_test` or `set_success`. 

I completely understand if this is outside the core project though. Incredible work regardless!

**Reply by acbart** [reply]
Perfectly fine to talk about it here! This is definitely Pedal oriented, even if it touches on an environment we're integrated with. Can you share the code you're working with? 

The default model we use in Pedal is that a submission is correct until a negative feedback is encountered (which can happen from student runtime errors, failing unit tests, etc). Separate from that is the idea of a score. Unless I am mistaken, a score of 1 will be set if there are no errors; otherwise the sum of the partial credits are used.

Most Feedback Functions can take a `score` parameter. These can either be floats (between 0 and 1) or strings like `"+10%"`. The `unit_test` function in particular has several ways to allocate partial credit depending on the quantity of unit tests that pass.

If you give me some more context of what you're doing, I can probably provide more specific help!

Thinking about BlockPy, it's also worth pointing out that the instructor interface can be a little confusing in how it reports scores, since it follows the student model that it never lowers a score - if you earned a 100% at some point, that value persists.

**Reply by jlehenbauer** [reply]
All really helpful information, thank you!

I'm definitely accomplishing something, but as you said, the BlockPy instructor side can be slightly opaque 😅😅

It may be worth noting that I'm doing the following in `On Run`, as `On Eval` doesn't seem to give feedback.

I'm grading a simple looping assignment like this:
```
from pedal import *

verify()

ast = parse_program()

# This works as expected
if len(ast.find_all('For')) < 2:
    gently("Please use `for` loops for each function")

# This complement appears as shown in the following screenshot, but doesn't affect the grade
if unit_test('sum_list', ([[1, 2, 3]], 6)):
    compliment("Good job adding up the elements in the list!", score="+33%")

# This will report a failing of the test, but not affect the grade (see second screenshot)
unit_test('find_greatest', ([[1, 2, 3]], 3), score = "+33%")
```

Even hardcoding something like `set_success(score = .8)` yields no changes.

![image](https://user-images.githubusercontent.com/3521602/151820307-fae10e69-4c36-406c-a9bb-d1e9ab239a95.png)

![image](https://user-images.githubusercontent.com/3521602/151820410-365318da-5aad-478e-91ed-63b3c4819605.png)

In addition, no matter what these scores are, the course screen always reports these as "Failed", even though "Correct" (Yes/No) indicates correctly whether they've passed all the tests.

![image](https://user-images.githubusercontent.com/3521602/151820652-8c064956-2079-4ef8-92d0-f024b3a022f5.png)

Again, I know all this may be more tied to BlockPy than the Petal project, so I understand if feedback is limited. I appreciate the listening ear regardless!

**Reply by jlehenbauer** [reply]
I've got some revisions:

In the manner you described, the "Feedback" score visible to the instructor only increases. I've been able to successfully increase this score from 0 to 100 using the `score` modifiers as intended!

This problem still remains though, and my only guess is that it may be related to not being connected to Canvas as an LMS?

> In addition, no matter what these scores are, the course screen always reports these as "Failed", even though "Correct" (Yes/No) indicates correctly whether they've passed all the tests.
> 
> ![image](https://user-images.githubusercontent.com/3521602/151820652-8c064956-2079-4ef8-92d0-f024b3a022f5.png)

**Comment by eryckpedro** [comment]
Hello. I'm currently trying to use Pedal to detect and count the number of occurrences of some code patterns I've manually encountered in CS1 student code, and I want to know more about what Pedal can do with `_var_` and `__expr__`. I'm going to show an example below:

One of the examples I've encountered is when students declare consecutive if statements with the **same** condition in both, but with **distinct** operations inside their blocks. A general example would be:

**example.py**: 
``` python
1. a = 1
2. b = 6
3. if a * b >= 6:
4.     print(a, b)
5. if a * b >= 6:
6.     print(a * b)
```

The current code with Pedal I'm trying to detect such pattern is:

**instructor.py**:
``` python
1. match = find_match("if __expr__:\n"
2.                    "    ___\n"
3.                    "if __expr__:\n"
4.                    "    ___", code) #here I've opened and read example.py beforehand and stored in code variable
5. if match:
6.     print("B12: Consecutive equal if statements with distinct operations in their blocks.")
```

What I'm wondering is if Pedal recognizes that, in **instructor.py**, both `__expr__` in lines 1 and 3 will match an exact expression (i.e. `a * b >= 6` in lines 3 and 5 in **example.py**). From my tests, if I change lines 3 and 5 (in **example.py**), the `find_match` **still** detects the pattern. Do I need to ensure it's the same expression in another way? If so, how can I do that? Does Pedal provide any way to do it? Thanks in advance.

PS: I'm quite unfamiliar with ASTs, I'm trying to use Pedal because CAIT seems very promising to find the patterns I've found manually.

**Reply by lukesg08** [reply]
For tree matching in find matches, we do not have the capability at the moment to test whether two subtrees are equal using \_\_expr__ (detailed a bit in the documentation I think), BUT, we can do something like this:

```python
match = find_match("if __expr1__:\n"
                             "    ___\n"
                             "if __expr2__:\n"
                             "    ___", code)
if find_match(match["__expr1__"].astNode, match["__expr2__"].astNode):
    print("B12: Consecutive equal if statements with distinct operations in their blocks.")
```
This functionality has not been added to the documentation yet. Basically, we find the match first, then we reference the match to find the two expressions we are denoting as "\_\_expr1__" and "\_\_expr2__", and find a match between the asts referenced by those two expressions!

Variables on the other hand, do have matching comparisons, so if you have two variable place holders, the matching will check that the names of the variables match with each other (detailed more in the documentation).

Also, this will only check if the two expressions (subtrees) are EXACTLY the same. We don't have a way to check whether two boolean expressions are logically the same.
